libnavigation-core/src/androidTest/java/com/mapbox/navigation/core/trip/service/ArtificialDriverTest.kt

```diff
@@ -0,0 +1,143 @@
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.MapboxNavigationProvider
+import com.mapbox.navigation.core.navigator.toFixLocation
+import com.mapbox.navigation.core.replay.history.ReplayEventUpdateLocation
+import com.mapbox.navigation.core.replay.history.mapToLocation
+import com.mapbox.navigation.core.replay.route.ReplayRouteMapper
+import com.mapbox.navigation.core.test.R
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.NavigationStatusOrigin
+import com.mapbox.navigator.NavigatorObserver
+import com.mapbox.navigator.RouteState
+import com.mapbox.navigator.SetRoutesReason
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert.assertTrue
+import org.junit.Ignore
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+@RunWith(AndroidJUnit4::class)
+class ArtificialDriverTest {
+
+    @Test
+    @Ignore("test sometimes fails because of https://mapbox.atlassian.net/browse/NN-418")
+    fun nativeNavigatorFollowsArtificialDriverWithoutReroutes() =
+        runBlocking<Unit>(Dispatchers.Main) {
+            withNavigators { mapboxNavigation, nativeNavigator ->
+                mapboxNavigation.historyRecorder.startRecording()
+                val testRoute = getTestRoute()
+                val events = createArtificialLocationUpdates(testRoute)
+                val setRoutesResult =
+                    nativeNavigator.setRoutes(testRoute, reason = SetRoutesReason.NEW_ROUTE)
+                assertTrue("result is $setRoutesResult", setRoutesResult.isValue)
+                val statusesTracking = async<List<NavigationStatus>> {
+                    nativeNavigator.collectStatuses(untilRouteState = RouteState.COMPLETE)
+                }
+
+                for (location in events.map { it.location.mapToLocation() }) {
+                    assertTrue(nativeNavigator.updateLocation(location.toFixLocation()))
+                }
+
+                val states = statusesTracking.await()
+                val historyFile = suspendCoroutine<String> { continuation ->
+                    mapboxNavigation.historyRecorder.stopRecording {
+                        continuation.resume(it ?: "null")
+                    }
+                }
+                val notTrackingStates = states.filter { it.routeState != RouteState.TRACKING }
```

Comment:
Is this the right check? I mean do we want to test that we only have tracking or that we don't have off_route? What about INITIALIZED state for example?

Reply:
I wrote this test to verify absence of reroutes when location updates are simulated by our artificial driver. That's why I check only tracking part. But I'm open to suggestions, if you think a different check will bring us more benefits, I'm happy to change the current one 🙂

Reply:
Oops, I thought I left a comment but I didn't. Judging by the name of the test, I'd do sth like:
```
val offRouteStates = states.filter { it.routeState == RouteState.OFF_ROUTE }
```

Reply:
Good idea! applied